gs_quant/timeseries/arima.py

```diff
@@ -0,0 +1,248 @@
+
+    Moving average components uses past forecast errors E(t). In other words,
+    X(t) can be thought of as a weighted moving average of the past forecast 
+    errors: X(t) = c + E(t) + W(1)*E(t-1) + ... + W(q)*E(t-q).
+    """
+
+    def __init__(self):
+        self.best_params = {}
+    
+
+    def _evaluate_arima_model(self, X: Union[pd.Series, pd.DataFrame], arima_order: Tuple[int, int, int], train_size: float, freq: str) -> Tuple[float, dict]:
+        train_size = int(len(X) * train_size)
+        train, test = X[0:train_size].astype(float), X[train_size:].astype(float)
+
+        model = ARIMA(train, order=arima_order, freq=freq)
+        model_fit = model.fit(disp=False, method='css', trend="nc")
+        ar_coef = model_fit.arparams
+        ma_coef = model_fit.maparams
+        resid = model_fit.resid
+
+        model_params = model_fit.params.to_dict()
+        const = model_params.get("const", 0)
+
+        # calculate test error
+        yhat = model_fit.forecast(len(test))[0]
+        error = mse(test, yhat)
+
+        return error, const, ar_coef, ma_coef, resid
+
+
+    def fit(self, X: Union[pd.Series, pd.DataFrame], train_size: float, p_vals: list = [0,1,2], d_vals: list=[0,1,2], q_vals: list=[0,1,2], freq: str=None) -> arima: 
+        """
+        Train a combination of ARIMA models. If pandas DataFrame, finds the 
+        best arima model parameters for each column. If pandas Series, finds 
+        the best arima model parameters for the series.
+
+        :param X: time series to be operated on; required parameter
+        :param train_size: between 0.0 and 1.0 and represents the proportion of the dataset to include in the train split
+        :p_vals: number of autoregressive terms to search; default is [0,1,2]
+        :d_vals: number of differences to search; default is [0,1,2]
+        :q_vals: number of lagged forecast to search; always [0,1,2]
+        :freq: frequency of time series, default is None
+        :return: self
+        """
+
+        if isinstance(X, pd.Series): X = X.to_frame()
+
+        if isinstance(X, pd.DataFrame):
+            for series_id in tqdm(X.columns):
+                series = X[series_id]
+                best_score, best_order, best_const, best_ar_coef, best_ma_coef, best_resid = float("inf"), None, None, None, None, None
+                for order in list(itertools.product(*[p_vals, d_vals, q_vals])):
+                    try:
+                        mse, const, ar_coef, ma_coef, resid = self._evaluate_arima_model(series, order, train_size, freq)
+                        if mse < best_score:
+                            best_score = mse
+                            best_order = order
+                            best_const = const
+                            best_ar_coef = ar_coef
+                            best_ma_coef = ma_coef
+                            best_resid = resid
+                    except Exception as e:
+                        print('   {}'.format(e))
```

Comment:
pls raise exception, remove print

Reply:
Certain combinations of (p, q, d) will raise the following exception: `Estimation requires the inclusion of least one AR term, MA term, a constant or an exogenous variable.`

Raise exception will then break the training loop. Maybe it's a better idea to just print the error and move on to the next combination of (p, q, d)?